scriptworker_client/src/scriptworker_client/aio.py

```diff
@@ -81,6 +89,163 @@
         return await coro
 
 
+# lockfile {{{1
+@asynccontextmanager
+async def lockfile(paths, name=None, attempts=10, sleep=30):
+    """Acquire a lockfile from among ``paths`` and yield the path.
+
+    If we want inter-process semaphores, we can use lock files rather than
+    ``asyncio.Semaphore``.
+
+    The reason we're using ``open(path, "x")`` rather than purely relying
+    on ``fcntl.lockf`` is that fcntl file locking doesn't lock the file
+    inside the current process, only for outside processes. If we don't
+    keep track of which lockfiles we've used in our async script, we can
+    easily blow away our own lock if we, say, used ``open(path, "w")`` or
+    ``open(path, "a") and then closed any of the open filehandles.
+
+    (See http://0pointer.de/blog/projects/locking.html for more details.)
+
+    Args:
+        paths (list): a list of path strings to use as lockfiles.
+        name (str, optional): a descriptive name for the process that needs
+            the lockfile, for logging purposes. Defaults to ``None``.
+        attempts (int, optional): the number of attempts to get a lockfile.
+            This means we attempt to get a lockfile from every path in ``paths``, ``attempts`` times. Defaults to 20.
+        sleep (int, optional): the number of seconds to sleep between attempts.
+            We sleep after attempting every path in ``paths``. Defaults to 30.
+
+    Yields:
+        str: the lockfile path acquired.
+
+    Raises:
+        LockfileError: if we've exhausted our attempts.
+
+    """
+    if name is not None:
+        acquired_msg = "Lockfile acquired for {} at %s".format(name)
+        wait_msg = "Couldn't get lock for {}; sleeping %s".format(name)
+        failed_msg = "Can't get lock for {} from paths %s after %s attempts".format(
+            name
+        )
+    else:
+        acquired_msg = "Lockfile acquired at %s"
+        wait_msg = "Couldn't get lock; sleeping %s"
+        failed_msg = "Can't get lock from paths %s after %s attempts"
+    for attempt in range(0, attempts):
+        for path in [item for item in random.sample(paths, len(paths))]:
+            try:
+                # Ensure the file doesn't exist, so we don't blow away
+                # our own lockfiles.
+                with open(path, "x") as fh:
+                    # Acquire an fcntl lock, in case other processes
+                    # use something other than ``lockfile`` to acquire
+                    # locks
+                    try:
+                        fcntl.lockf(fh, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                        log.debug(acquired_msg, path)
+                        yield path
+                        return
```

Comment:
yield followed by return is a bit of a code smell for me - it effectively turns this into a regular function. The alternative would be dropping the context manager, and just straight assigning this to a value, I think? It's possible I'm missing one of the benefits of using a context manager for this.

As I wrote the above I think I realized this _does_ actually do something on __aexit__, namely: remove the lockfile. That appears to be in `finally` to clean it up for both the success and failure case. If all of that's correct, I'm fine with this as is, but could you add a brief explanatory comment for posterity, as it's not obvious?

Reply:
Yes, this is to avoid the `LockfileError` below and also to hit the `finally` block. Sure, I can comment.